Implement BackingStoreSkia

donny-dont opened the pull request:
#### f8ceb0c9e43e72c947be2f388034a14bb7df27f6
<pre>
Implement BackingStoreSkia
<a href="https://bugs.webkit.org/show_bug.cgi?id=273593">https://bugs.webkit.org/show_bug.cgi?id=273593</a>

Reviewed by Carlos Garcia Campos.

Fill in the implementation of `BackingStore` when `USE(SKIA)` is `ON`. Follow
along with the behavior of the Cairo implementation.

* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:

Canonical link: <a href="https://commits.webkit.org/278877@main">https://commits.webkit.org/278877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab6559d8c121d265244a21084f207c8c03410ed3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51800 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4164 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/55066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/2492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2197 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/55066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53899 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28733 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/44686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/55066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47982 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/2038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56659 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26921 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/44782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29055 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7570 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
```diff
@@ -46,17 +58,61 @@ BackingStore::~BackingStore() = default;
+        m_scrollSurface = createCairoImageSurfaceWithFastMalloc(m_size, m_deviceScaleFactor);
+
+    //copyRectFromOneSurfaceToAnother(m_surface.get(), m_scrollSurface.get(), scrollOffset, targetRect);
+    //copyRectFromOneSurfaceToAnother(m_scrollSurface.get(), m_surface.get(), { }, targetRect);
```
@carlosgcampos any pointers on what should go here? Everything else follows along with the implementation in `BackingStoreCairo.cpp`.

donny-dont commented on Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
```diff
@@ -29,15 +29,39 @@
+        return;
+
+    destination->writePixels(bitmap, rect.x(), rect.y());
+}
```
Think I translated from Cairo wrong with this one but getting closer. @carlosgcampos not sure if you have any thoughts on how this should be implemented here.

carlosgcampos commented on Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
```diff
@@ -29,34 +29,93 @@
 namespace WebKit {
 
-static const Seconds s_scrollHysteresisDuration { 300_ms };
+static sk_sp<SkSurface> createSkiaImageSurfaceWithFastMalloc(const WebCore::IntSize& size, double deviceScaleFactor)
```
Note that we are already using a custom allocator for skia that uses bmalloc, so you don't need to explicitly alloc data.

carlosgcampos commented on Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
```diff
@@ -29,34 +29,93 @@
+    scaledSize.scale(deviceScaleFactor);
+    auto info = SkImageInfo::MakeN32Premul(scaledSize.width(), scaledSize.height(), SkColorSpace::MakeSRGB());
+    auto* pixels = fastZeroedMalloc(scaledSize.height() * info.minRowBytes());
+    return SkSurfaces::WrapPixels(info, pixels, info.minRowBytes(), nullptr);
```
The data is leaked. Since skia will already use bmalloc, I think it would be simpler to use `SkSurfaces::Raster` instead of `SkSurfaces::WrappPixels`, and skia will own the pixels.

carlosgcampos commented on Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:
```diff
@@ -29,34 +29,93 @@
+    // Paint all update rects.
+    WebCore::IntPoint updateRectLocation = updateInfo.updateRectBounds.location();
+
+    SkCanvas* canvas = m_surface->getCanvas();
+    if (!canvas)
+        return;
```
I think you should also nullcheck m_surface.

donny-dont replied:
Added a `RELEASE_ASSERT` that checks the `m_surface`

donny-dont committed:
Implement BackingStoreSkia
https://bugs.webkit.org/show_bug.cgi?id=273593

Reviewed by Carlos Garcia Campos.

Fill in the implementation of `BackingStore` when `USE(SKIA)` is `ON`. Follow
along with the behavior of the Cairo implementation.

* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/UIProcess/BackingStore.h:
* Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp:

Canonical link: https://commits.webkit.org/278877@main
